src/keystore.ts

```diff
@@ -0,0 +1,131 @@
+      case "scrypt":
+        return scrypt(password, args as ScryptParams);
+      case "pbkdf2":
+        return PBKDF2(password, args as PBKDF2Params);
+      default:
+        throw new Error("Unsupported crypto function");
+    }
+  }
+
+  constructor(keystore?: IKeystoreParams){
+    if(keystore){
+      this.crypto = new KeystoreCrypto(keystore.crypto);
+      this.pubkey = keystore.pubkey || "";
+      this.path = keystore.path || "";
+      this.uuid = keystore.uuid || uuid.v4();
+      this.version = keystore.version || 4;
+    }
+  }
+
+  public static fromJson(json: Record<string, any>): Keystore {
+    const jsonObj = json as IKeystoreParams;
+    const keystore: IKeystoreParams = {
+      crypto: KeystoreCrypto.fromJson(jsonObj.crypto || {}),
+      path: jsonObj.path,
+      pubkey: jsonObj.pubkey,
+      uuid: jsonObj.uuid,
+      version: jsonObj.version,
+    };
+    return new Keystore(keystore);
+  }
+
+  public static encrypt(secret: bytes, password: string, path = "", kdfSalt: bytes = random.randomBuffer(32), aesIv: bytes = random.randomBuffer(16)): IKeystore {
+    
+    const keystore = new this({ 
+      path: path,
+      pubkey: PublicKey.fromPrivateKey(PrivateKey.fromBytes(secret)).toHexString().replace("0x", ""),
+      crypto: {
+        kdf: {
+          params: {
+            salt: kdfSalt
+          }
+        },
+        cipher: {
+          params: {
+            iv: aesIv
+          }
+        }
+      }
+    });
+
+    const decryptionKey: bytes = keystore.kdf(password, keystore.crypto.kdf.params);
+    const cipher = AES_128_CTR(decryptionKey.slice(0, 16), keystore.crypto.cipher.params.iv);
+
+    let encryptedSecret = cipher.update(secret);
+    encryptedSecret = Buffer.concat([encryptedSecret, cipher.final()]);
+
+    keystore.crypto.cipher.message = Buffer.from(encryptedSecret);
+    keystore.crypto.checksum.message = SHA256(Buffer.concat([decryptionKey.slice(16, 32), keystore.crypto.cipher.message]));
+    
+    return keystore;
+  }
+
+  public verifyPassword(password: string): boolean {
+    const decryptionKey: bytes = this.kdf(password, this.crypto.kdf.params);
+    if( SHA256(Buffer.concat([decryptionKey.slice(16, 32), this.crypto.cipher.message])).compare(this.crypto.checksum.message) === 0){
+      // Password matches
+      return true;
+    }
+    return false;
+  }
+
+  public decrypt(password: string): Buffer {
+
+    if(this.verifyPassword(password) === false){
+      throw new Error("Invalid password");
+    }
+    const decryptionKey: bytes = this.kdf(password, this.crypto.kdf.params);
+    const cipher = AES_128_CTR(decryptionKey.slice(0, 16), this.crypto.cipher.params.iv);
+    let decryptedSecret = cipher.update(this.crypto.cipher.message);
+    decryptedSecret = Buffer.concat([decryptedSecret, cipher.final()]);
+    return decryptedSecret;
+  }
+
+}
+
+export class Pbkdf2Keystore extends Keystore {
+  constructor(keystore?: IKeystoreParams) {
+    super(keystore);
```

Comment:
modify keystore params with set values before passing to constructor instead of modifying object from outside